async/async.go

```diff
@@ -0,0 +1,372 @@
+			Subject:     to.Ptr(string(taskID)),
+		},
+	}
+
+	if _, err := h.pubsub.PublishEvents(ctx, events, nil); err != nil {
+		return fmt.Errorf("failed to publish event: %w", err)
+	}
+
+	h.logger.Info("created report", zap.String("task_id", string(taskID)))
+
+	return nil
+}
+
+func (h *Handler) downloadBlob(ctx context.Context, srcURI, destPath string) error {
+	// Make sure the destination exists
+	if err := os.MkdirAll(filepath.Dir(destPath), 0600); err != nil {
+		return fmt.Errorf("failed to create directory to download blob to: %w", err)
+	}
+
+	destF, err := os.Create(destPath)
+	if err != nil {
+		return fmt.Errorf("failed to create dest file: %w", err)
+	}
+	defer destF.Close() // Best-effort in case something fails
+
+	br, err := h.blob.ReadBlob(ctx, srcURI)
+	if err != nil {
+		return fmt.Errorf("failed to read raw portfolio: %w", err)
+	}
+	defer br.Close() // Best-effort in case something fails
+
+	if _, err := io.Copy(destF, br); err != nil {
+		return fmt.Errorf("failed to load raw portfolio: %w", err)
+	}
+
+	if err := br.Close(); err != nil {
+		return fmt.Errorf("failed to close blob reader: %w", err)
+	}
+
+	if err := destF.Close(); err != nil {
+		return fmt.Errorf("failed to close dest file: %w", err)
+	}
+
+	return nil
+}
+
+func (h *Handler) uploadDirectory(ctx context.Context, dirPath, container string) ([]*task.AnalysisArtifact, error) {
+	base := filepath.Base(dirPath)
+
+	var artifacts []*task.AnalysisArtifact
+	err := filepath.WalkDir(dirPath, func(path string, info fs.DirEntry, err error) error {
+		if info.IsDir() {
+			return nil
+		}
+
+		// This is a file, let's upload it to the container
+		uri := blob.Join(h.blob.Scheme(), container, base, strings.TrimPrefix(path, dirPath+"/"))
+		if err := h.uploadBlob(ctx, path, uri); err != nil {
+			return fmt.Errorf("failed to upload blob: %w", err)
+		}
+
+		fn := filepath.Base(path)
+		// Returns pacta.FileType_UNKNOWN for unrecognized extensions, which we'll serve as binary blobs.
+		ft := fileTypeFromExt(filepath.Ext(fn))
+		if ft == pacta.FileType_UNKNOWN {
+			h.logger.Error("unhandled file extension", zap.String("dir", dirPath), zap.String("file_ext", filepath.Ext(fn)))
+		}
+		artifacts = append(artifacts, &task.AnalysisArtifact{
+			BlobURI:  pacta.BlobURI(uri),
+			FileName: fn,
+			FileType: ft,
+		})
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error while walking dir/uploading blobs: %w", err)
+	}
+	return artifacts, nil
+}
+
+func fileTypeFromExt(ext string) pacta.FileType {
+	switch ext {
+	case ".csv":
+		return pacta.FileType_CSV
+	case ".yaml":
+		return pacta.FileType_YAML
+	case ".zip":
+		return pacta.FileType_ZIP
+	case ".html":
+		return pacta.FileType_HTML
+	case ".json":
+		return pacta.FileType_JSON
+	case ".txt":
+		return pacta.FileType_TEXT
+	case ".css":
+		return pacta.FileType_CSS
+	case ".js":
+		return pacta.FileType_JS
+	case ".ttf":
+		return pacta.FileType_TTF
+	default:
+		return pacta.FileType_UNKNOWN
```

Comment:
Can we add a log here for missing cases?

Reply:
Indeed we do! The method is kept simple for separation of concern purposes, but we do indeed log an error if the extension is `UNKNOWN`